lib/pbio/drv/usb/usb_ev3.c

```diff
@@ -52,500 +28,823 @@
+}
 
-// Global flag indicating that a USB configuration has been set.
-static volatile bool g_bUSBConfigured = false;
-static bool g_bUSBSerialConfigured = false;
 
-static void GetLineCoding(tLineCoding *psLineCoding) {
-    //
-    // Get the current line coding set in the UART.
-    //
-    psLineCoding->ulRate = 115200;
-    psLineCoding->ucDatabits = 8;
-    psLineCoding->ucParity = USB_CDC_PARITY_NONE;
-    psLineCoding->ucStop = USB_CDC_STOP_BITS_1;
+// Fill in the CPPI DMA descriptor to send a packet
+static void usb_setup_tx_dma_desc(int tx_type, void *buf, uint32_t buf_len) {
+    cppi_descriptors[tx_type] = (usb_cppi_hpd) {
+        .word0 = {
+            .hostPktType = 0x10,
+            .pktLength = buf_len,
+        },
+        .word1 = {
+            .srcPrtNum = 1,             // port is EP1
+        },
+        .word2 = {
+            .pktType = 5,               // USB packet type
+            .pktRetQueue = TX_COMPQ1,
+        },
+        .buf_len = buf_len,
+        .buf_ptr = buf,
+        .next_desc_ptr = 0,
+        .orig_buf_len = buf_len,
+        .orig_buf_ptr = buf,
+    };
+
+    __asm__ volatile ("" ::: "memory");
+
+    HWREG(USB_0_OTGBASE + CPDMA_QUEUE_REGISTER_D + TX_SUBMITQ1 * 16) =
+        (uint32_t)(&cppi_descriptors[tx_type]) | 2;
+}
+
+// Helper function for dividing up buffers for EP0 and feeding the FIFO
+static void usb_setup_send_chunk() {
+    unsigned int this_chunk_sz = setup_data_to_send_sz;
+    if (this_chunk_sz > EP0_BUF_SZ) {
+        this_chunk_sz = EP0_BUF_SZ;
+    }
 
+    // 4-byte-at-a-time copy
+    unsigned int this_chunk_sz_words = this_chunk_sz / 4;
+    for (unsigned int i = 0; i < this_chunk_sz_words; i++) {
+        HWREG(USB0_BASE + USB_O_FIFO0) = setup_data_to_send[i];
+    }
+    setup_data_to_send += this_chunk_sz_words;
+
+    // Copy remainder
+    for (unsigned int i = 0; i < this_chunk_sz % 4; i++) {
+        HWREGB(USB0_BASE + USB_O_FIFO0) = ((const uint8_t *)setup_data_to_send)[i];
+    }
+
+    setup_data_to_send_sz -= this_chunk_sz;
 }
 
-// *****************************************************************************
-//
-// Handles CDC driver notifications related to the receive channel (data from
-// the USB host).
-//
-// \param ulCBData is the client-supplied callback data value for this channel.
-// \param ulEvent identifies the event we are being notified about.
-// \param ulMsgValue is an event-specific value.
-// \param pvMsgData is an event-specific pointer.
-//
-// This function is called by the CDC driver to notify us of any events
-// related to operation of the receive data channel (the OUT channel carrying
-// data from the USB host).
-//
-// \return The return value is event-specific.
-//
-// *****************************************************************************
-unsigned int RxHandler(void *pvCBData, unsigned int ulEvent, unsigned int ulMsgValue, void *pvMsgData) {
+static void usb_device_intr() {
```

Comment:
```suggestion
static void usb_device_intr(void) {
```

And other places as well.